app/models/concerns/folio/console/clonable.rb

```diff
@@ -0,0 +1,152 @@
+    log(I18n.t("cloning.finished"))
+    clone
+rescue => e
+  log(I18n.t("cloning.error", message: e.message), :error)
+  log(e.backtrace.first(5).join("\n"), :error)
+  raise
+  end
+
+    private
+      def reset_clone_attributes(clone)
+        self.class.reset_attributes.each do |attr|
+          clone[attr] = nil if clone.has_attribute?(attr)
+        end
+      end
+
+      def copy_references(cloned)
+        return unless self.class.reference_associations.present?
+
+        self.class.reference_associations.each do |assoc|
+          cloned.public_send("#{assoc}=", public_send(assoc))
+        end
+      end
+
+      def duplicate_nested_records(cloned)
+        return unless self.class.duplicated_associations.present?
+
+        self.class.duplicated_associations.each do |assoc|
+          cloned.public_send("#{assoc}=", clone_associated_records(assoc))
+        end
+      end
+
+      def clone_associated_records(association)
+        originals = public_send(association)
+        originals = [originals] unless originals.is_a?(ActiveRecord::Relation)
+
+        clones = originals.map do |orig|
+          clone = orig.deep_dup
+
+          if orig.class.reflect_on_all_associations.present?
+            orig.class.reflect_on_all_associations.each do |association|
+              clone_association(orig, clone, association)
+            end
+          end
+
+          clone
+        end
+
+        originals.is_a?(ActiveRecord::Relation) ? clones : clones.first
+      end
+
+      def clone_association(original, clone, association)
+        if association.macro == :has_many
+          clone_has_many_association(original, clone, association)
+        else
+          clone_single_association(original, clone, association)
+        end
+      end
+
+      def clone_has_many_association(original, clone, association)
+        return if [:files, :placements].include?(association.name)
+        associated_records = original.public_send(association.name)
+        associated_records = associated_records.map { |r| r.deep_dup } unless self.class.reference_associations.include?(association.name)
+        clone.association(association.name).build(associated_records.map(&:attributes))
+      end
+
+      def clone_single_association(original, clone, association)
+        associated_record = original.public_send(association.name)
+        return unless associated_record
+
+        associated_record_dup = self.class.reference_associations.include?(association.name) ?
+                              associated_record :
+                              associated_record.deep_dup
+
+        clone.public_send("#{association.name}=", associated_record_dup)
+      end
+
+      def log(message, level = :info)
+        if Rails.env.development? && Rails.logger
+          Rails.logger.tagged("CLONING") do
+            Rails.logger.public_send(level, message)
+          end
+        else
+          puts "[CLONING] #{message}"
+        end
+      end
+end
```

Comment:
Moc se mi nelibi, ze je to cele jako concern na model. Pridava to strasne moc metod vcetne `log` a nejakych dalsich, ktere jsou dost obecne nazvane.

Predstovoval bych si to jako concern, ktery prida neco jako `self.is_clonable?` true / false a `clone` metody. `clone` by pak uz melo pracovat skrze vlastni tridu podobne jako napr. https://github.com/sinfin/folio/blob/v6/app/lib/folio/transportable/exporter.rb a jen vratit novy neulozeny zaznam.

Zaroven by to melo fungovat vice magicky a pro vsechny zaznamy (dejme to na `Folio::ApplicationRecord` a cele to v `default_actions` zapinejme pres Rails app config klic), lze pouzit `reflect_on_all_associations` a spis nez whitelistovat naopak blacklistovat per-model pres nejakou `self.metodu`, co nechci kopirovat. A pak pridat jeste nejaky volitelny callback na model zase jako tridni metodu.